Testing for the Secondary module

awwad opened the pull request:
Add module test_secondary.py to test the secondary.py module in the reference implementation. This is based on test_primary.py.

Two clients are used in the various tests, one of them distinguished by the fact that there is no Director metadata for it.

awwad committed:
Add test_secondary.py, based on test_primary.py

to test the reference implementation secondary.py module.

awwad committed:
DOC: Improve documentation in test_secondary

by adding some docstrings and adding comments around a few
potentially confusing actions.

awwad committed:
In test_secondary, move client directory creation to setUpModule

from test_01_init. It's part of the setup for all the tests, and
is not part of testing the constructor, so setUpModule is where
it belongs.

awwad committed:
Add testing for a second Secondary client in test_secondary

which will have different data to process. In particular, it will
be missing Director metadata.

awwad committed:
Rewrite Secondary tests for three clients and situations

one client with normal update data, one with no updates assigned
to it, and one with no Director metadata available to it at all.

Make the code less redundant with some loops.

Switch tuf.conf.repository_directory when switching among clients.

Add check for updated fileinfo data.

awwad committed:
Test validate_image method for Secondary

awwad committed:
Remove unnecessary import (import copy) in test_secondary

awwad committed:
Treat Secondary.validated_targets_for_this_ecu as a class field,

initializing it in Secondary.__init__() rather than only after
validation of metadata is attempted.

This allows easier testing and is also more intuitive.

vladimir-v-diaz commented on tests/test_secondary.py:
```diff
@@ -0,0 +1,721 @@
+"""
+<Program Name>
```
I know that our code style guidelines only require the `<Program Name>` and `<Purpose>` fields, but we should consider including the `<Copyright>` label to reference the LICENSE file.

awwad replied:
Adding, but it would be good to decide in a central location if this is something we should be doing.

vladimir-v-diaz commented on tests/test_secondary.py:
```diff
@@ -0,0 +1,721 @@
+key_directortargets_pub = None
+clock = None
+
+# Set starting firmware fileinfo (that this ECU had coming from the factory)
```
So far you've done a good job explaining why certain variables are needed.  It's not entirely clear to me why this firmware fileinfo is set here.

awwad replied:
Adding note
```
# It will serve as the initial firmware state for the Secondary clients.
```

vladimir-v-diaz commented on tests/test_secondary.py:
```diff
@@ -0,0 +1,721 @@
+# For temporary convenience:
+import demo # for generate_key, import_public_key, import_private_key
+
+TEST_DATA_DIR = os.path.join(uptane.WORKING_DIR, 'tests', 'test_data')
```
Why isn't the `tempfile` module used instead?
https://docs.python.org/2/library/tempfile.html

awwad replied:
I haven't considered it yet. This works and is what is already done in test_primary. Should consider it later.

vladimir-v-diaz commented on tests/test_secondary.py:
```diff
@@ -0,0 +1,721 @@
+  """
+  This is run once for the full module, before all tests.
+  It prepares some globals, including two for Secondary ECU client instances.
+  When finished, it will also start up an OEM Repository Server,
```
A nitpick, but you might want to add "When this module is finished, it will ..."  At first, I thought you meant when this function ends its execution.

awwad replied:
Ah, that's text to be stripped and I've removed it in a commit on another branch to be PR'd soon. (I didn't catch it here in time and now it's already there. :P)

vladimir-v-diaz commented on tests/test_secondary.py:
```diff
@@ -0,0 +1,721 @@
+  Unit testing for uptane/clients/secondary.py
+  Much of this is copied from test_primary.py and then modified slightly
+"""
+from __future__ import print_function
```
Is there a reason why anything is printed to the console for this unit test (and others?)  Ideally, only message about test failures should appear on the console.

awwad replied:
Debugging reasons, but you're right: I shouldn't spam. I'll strip it out here and try to remember to do likewise in the other test modules.

vladimir-v-diaz commented on tests/test_secondary.py:
```diff
@@ -0,0 +1,721 @@
+    global secondary_instance_3
+
+    # TODO: Test with invalid pinning file
+    # TODO: Test with pinning file lacking a Director repo.
```
I don't know for certain, but make sure this comment is still true.  The pull request summary mentions that a client is tested that lacks Director metadata.  Related?

awwad replied:
It's slightly different: Director metadata being missing (tested already) vs pinning file failing to list a Director repository.

vladimir-v-diaz commented on tests/test_secondary.py:
```diff
@@ -0,0 +1,721 @@
+          director_public_key=key_directortargets_pub,
+          partial_verifying=False)
+
+    # Invalid director_public_key:
```
What about bad arguments for the others, like full_client_dir, director_repo_name, etc.?

awwad replied:
I'll add those two. The others are all addressed, I think.

vladimir-v-diaz commented on tests/test_secondary.py:
```diff
@@ -0,0 +1,721 @@
+
+    # Invalid director_public_key:
+    with self.assertRaises(tuf.FormatError):
+      s = secondary.Secondary(
```
Is it necessary to assign the secondary.Secondary instances to `s`?

awwad replied:
I suppose it doesn't matter, since these are all expected to raise exceptions. I'll remove the assignment in the next PR to avoid conflicts.

vladimir-v-diaz commented on tests/test_secondary.py:
```diff
@@ -0,0 +1,721 @@
+          time=clock,
+          timeserver_public_key=key_timeserver_pub,
+          firmware_fileinfo=factory_firmware_fileinfo,
+          director_public_key=key_directortargets_pub,
```
Instead of using two parameters, perhaps make it so that specifying a public director key is understood to mean that partial_verification is in effect.

awwad replied:
Hesitating. Explicit is good, but you're probably right. In any case, the partial verification functionality is not really tested. I'll probably leave this for when that's addressed in the future.

vladimir-v-diaz commented on tests/test_secondary.py:
```diff
@@ -0,0 +1,721 @@
+
+  # Generate a trusted initial time for the Secondary.
+  clock = tuf.formats.unix_timestamp_to_datetime(int(time.time()))
+  clock = clock.isoformat() + 'Z'
```
This variable name is a bit strange.  Seems like it should use the word `time`.

awwad replied:
Sure, I'll call it initial_time. To avoid conflicts, I'll make this change in the next PR, which already changes these lines a bit.

vladimir-v-diaz commented on tests/test_secondary.py:
```diff
@@ -0,0 +1,721 @@
+    # Try initializing three Secondaries, expecting the three calls to work.
+    # Save the instances for future tests as module variables (global keyword
+    # used above), to save time and code.
+    # TODO: Stick TEST_PINNING_FNAME in the right place.
```
Is this still an actual TODO?

vladimir-v-diaz commented on tests/test_secondary.py:
```diff
@@ -0,0 +1,721 @@
+      time_attestation = asn1_codec.convert_signed_metadata_to_der(
+          original_time_attestation, private_key=key_timeserver_pri, resign=True)
+
+    secondary_instance_1.validate_time_attestation(time_attestation)
```
I guess nothing happens if `time_attestation` is valid?  Might be clearer to assert that None is raised or say so in a comment.

awwad replied:
Noting in a comment.

vladimir-v-diaz commented on tests/test_secondary.py:
```diff
@@ -0,0 +1,721 @@
+      secondary_instance_1.validate_time_attestation(time_attestation__wrongnonce)
+
+
+    # TODO: Consider other tests here.
```
If not all major functionality is tested, we should ideally aim for at least 95% coverage.  We can do this incrementally.

awwad replied:
Yes.

vladimir-v-diaz commented on tests/test_secondary.py:
```diff
@@ -0,0 +1,721 @@
+
+    # TODO: Check some values from the ECU Manifest
+    # TODO: Check some values from the ECU Manifest
+    # TODO: Check some values from the ECU Manifest
```
Why three times? :D

awwad replied:
Already removed in another branch. Wanted to make sure I didn't miss it later.

awwad committed:
DOC: Add License info to test_secondary

awwad committed:
DOC: Explain factory_firmware_fileinfo in test_secondary

awwad committed:
Remove 'print's from test_secondary

awwad committed:
DOC: Note that a line in test_secondary succeeds silently

awwad committed:
Test more __init__ arguments for Secondary (test_secondary)

awwad committed:
DOC: Strip no-longer-relevant TODOs in test modules

awwad commented:
I think I've now addressed all your comments. Look OK now? (BTW, next PR for test_secondary will be to [this branch](https://github.com/awwad/uptane/compare/additional_testing...awwad:test_module_revisions?expand=1))

vladimir-v-diaz commented:
> (BTW, next PR for test_secondary will be to this branch)

Okay, just let me know when it's ready to ready review by assigning me as a reviewer.

I'll take a look at your follow-up commits to this pull request and let you know if it's okay.

vladimir-v-diaz commented:
This is good to merge when you're ready.
